Capture output of `eval` plugin and some girls scout changes

dschrempf opened the pull request:
This PR contains some girl-scout changes as well as changes to how the `eval` plugin treats `stdout` and `stderr`.

#### Eval plugin

We now capture output to `stdout` and `stderr` induced as a possible side effect by the statement being evaluated. This is fragile because the output may be scrambled in a concurrent setting when HLS is writing to one of these file handles from a different thread.

Fixes #1977, most likely also #4390 (although this is debatable, we could do something more sophisticated).

```haskell
-- >>> putStrLn "Hello"
-- Hello

-- >>> hPutStrLn stderr "Hello"
-- Hello

-- >>> putStrLn "Hello," >> pure "World!"
-- Hello,
-- "World!"
```

#### Girl scout changes

- Docs: Fix some links in `eval` plugin in-source documentation
- Clarify documentation of expected test failure
- Update Nix Flake (Darwin issues seem to be solved)
- Update some tests and avoid mentioning GHC 9.2
- Remove comment referring to GHC 9.2 and fix code

dschrempf committed:
Remove comment referring to GHC 9.2 and fix code

dschrempf committed:
Update some tests and avoid mentioning GHC 9.2

dschrempf committed:
Clarify documentation of expected test failure

dschrempf committed:
Docs: Fix some links in `eval` plugin in-source documentation

dschrempf committed:
eval: Capture `stdout` and `stderr`

- Redirect `stdout` and `stderr` to a temporary file
- Read back the file

fendor reviewed (approved):
Needs a proper test in `hls-eval-plugin`, but otherwise looks good to me, thank you for the clean up! Feel free to merge once you've added a test :)

dschrempf commented on plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Handlers.hs:
```diff
@@ -426,11 +428,18 @@ A, possibly multi line, error is returned for a wrong declaration, directive or
 Some flags have not been recognized: -XNonExistent
 
 >>> cls C
-Variable not in scope: cls :: t0 -> t
-Data constructor not in scope: C
+WAS Variable not in scope: cls :: t0 -> t
+WAS Data constructor not in scope: C
+NOW Illegal term-level use of the class `C'
+NOW   defined at <interactive>:1:2
+NOW In the first argument of `cls', namely `C'
+NOW In the expression: cls C
+NOW In an equation for `it_a5Zks': it_a5Zks = cls C
+NOW Variable not in scope: cls :: t0_a5ZlS[tau:1] -> t1_a5ZlU[tau:1]
```
I didn't realize the `eval` plugin changed some other results. Thanks for picking that up. I will investigate.

dschrempf replied:
@fendor 

But also everybody else if they have an opinion or a fact to add, please go ahead!

theGhostJW commented:
This could be an interesting surprise for some users. I often eval expressions that produce > 1000 lines of IO. In my case detailed pretty printed logs. If this was to suddenly be rendered in the code editor I'd need to change my workflow. If i wasn't expecting it it would be quite confusing.

dschrempf commented:
Yes, if this is your workflow. In theory, we could limit the amount of output. The underlying library redirects output to a temporary file and then reads this temporary file. We could amend the function so that it only takes the first N lines of this file (or bytes, or whatever).

dschrempf committed:
fourmolu: add missing `lsp` version bound

dschrempf committed:
eval: add missing `lsp` version bound

dschrempf commented:
Okay, finally some updates on this one.

- I had to revamp the method how we capture `stdout` and `stderr`. We now inject a setup and teardown function into the GHCi call. These calls setup the `stdout` and `stderr` capture, and also make sure the handles are closed afterwards. This is further ensured using `bracket` functions (new dependency on `exceptions`).
- I have changed the existing IO test to exercise the capture of `stdout` (instead of testing that it is _not_ captured).
- I have added another test that exercises capture of `stderr`.

I think we can merge this now, but waiting for your OK @fendor @xsebek.

dschrempf commented:
(I still have to refresh the in-source `eval` examples).

dschrempf committed:
eval: add short documentation to functions

xsebek commented on plugins/hls-eval-plugin/src/Ide/Plugin/Eval/Code.hs:
```diff
@@ -79,20 +92,115 @@ asStmts (Example e _ _) = NE.toList e
+    [ "(__hls_captureHandle, __hls_savedStdout, __hls_savedStderr) <- do {"
+    , "  __hls_h <- System.IO.openFile", show outputTemp, "System.IO.WriteMode;"
+    , "  System.IO.hSetBuffering __hls_h System.IO.LineBuffering;"
+    , "  __hls_o <- GHC.IO.Handle.hDuplicate System.IO.stdout;"
+    , "  __hls_e <- GHC.IO.Handle.hDuplicate System.IO.stderr;"
```
Could you please check if it is also possible to redirect standard input? The [docs](https://haskell-language-server.readthedocs.io/en/latest/features.html#evaluation-code-snippets-in-comments) mention it is currently broken:
> Standard input is shared with HLS, so e.g. [getLine breaks the connection to server](https://github.com/haskell/haskell-language-server/issues/2913).

But I guess fewer people run into that issue, while (`pretty`)`print` is very common. 🙂

dschrempf committed:
eval: avoid usage of term "test"; instead use "evalExpr"

xsebek reviewed (approved):
I tested it on the example from https://github.com/haskell/haskell-language-server/issues/4410 and it now works with just `putStrLn`, great! 🚀

dschrempf committed:
eval: fix multiline outputs

dschrempf committed:
eval: update readme; clean code; fix some bugs

- Correctly treat empty output or result when combining both
- Do not handle blank lines in any special way
- Update Readme of `hls-eval-plugin`

dschrempf commented:
I retract my statement about LGTM; there are still minor issues to fix (<BLANKLINE>?).

Also, some lenses get out of date (not sure if this is a problem of my editor setup, or a more general problem with HLS).

dschrempf commented:
OK, I gave it another go.

I fixed one more bug, improved large parts of the code (it is piece of art, but still a mess), and updated the Readme.

I think it is good to merge now. @xsebek, I didn't implement your feature request (I ran out of time); I think it can easily go in a separate PR.

Let's discuss tomorrow.
